File: 160-intersection-of-two-linked-lists/160-intersection-of-two-linked-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        map<ListNode *,int> tree;
        ListNode *nodeA,*nodeB;
        nodeA=headA;
        while(nodeA)
        {
            tree.insert(pair<ListNode *,int> (nodeA,nodeA->val));
            nodeA=nodeA->next;
        }
        nodeB=headB;
        map<ListNode *,int>::iterator i;
        while(nodeB)
        {
            i=tree.find(nodeB);
            if(i!=tree.end()) break;
            nodeB=nodeB->next;
        }
        if(nodeB==NULL) return NULL;
        return (*i).first;
    }
};

File: 160-intersection-of-two-linked-lists/160-intersection-of-two-linked-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        ListNode startA,startB; 
        startA=headA; startB=headB;
            while(true)
            {
                if(startA==startB) return startA;
                startA=startA.next;
                startB=startB.next;
                if(startA==null && startB==null) return null;
                if(startA==null) startA=headA; 
                if(startB==null) startB=headB;
                
            }
       
    }
}